In [468]:
import pandas as pd
import math

In [469]:
df = pd.read_excel('3.17.Confidence-intervals-exercise.xlsx')
df.drop(df.index[[0,1,2]], inplace=True)
df.drop(['Unnamed: 0', 'Unnamed: 12', 'Unnamed: 13'], inplace=True, axis=1)
df.rename(columns={'Inferential statistics. Confidence intervals': 'InvoiceNo' }, inplace=True)
df.rename(columns={'Unnamed: 2': 'Date' }, inplace=True)
df.rename(columns={'Unnamed: 3': 'Country' }, inplace=True)
df.rename(columns={'Unnamed: 4': 'ProductID' }, inplace=True)
df.rename(columns={'Unnamed: 5': 'Shop' }, inplace=True)
df.rename(columns={'Unnamed: 6': 'Gender' }, inplace=True)
df.rename(columns={'Unnamed: 7': 'Size (US)' }, inplace=True)
df.rename(columns={'Unnamed: 8': 'Size (Europe)' }, inplace=True)
df.rename(columns={'Unnamed: 9': 'Size (UK)' }, inplace=True)
df.rename(columns={'Unnamed: 10': 'UnitPrice' }, inplace=True)
df.rename(columns={'Unnamed: 11': 'Discount' }, inplace=True)
df.rename(columns={'Unnamed: 14': 'SalePrice' }, inplace=True)
df.head(4)

InvoiceNo                 Date         Country ProductID  Shop  Gender  \
3     52389  2014-01-01 01:06:24  United Kingdom      2152   UK2    Male   
4     52390  2014-01-01 03:44:48   United States      2230  US15    Male   
5     52391  2014-01-01 09:01:36          Canada      2160  CAN7    Male   
6     52392  2014-01-01 09:16:00   United States      2234   US6  Female   

  Size (US) Size (Europe) Size (UK) UnitPrice Discount SalePrice  
3        11            44      10.5       159        0       159  
4      11.5         44-45        11       199      0.2     159.2  
5       9.5         42-43         9       149      0.2     119.2  
6       9.5            40       7.5       159        0       159

In [470]:
df.tail(5)

InvoiceNo                 Date         Country ProductID  Shop  Gender  \
14965     65773  2016-12-31 13:35:12  United Kingdom      2154   UK2    Male   
14966     65774  2016-12-31 14:32:48   United States      2181  US12  Female   
14967     65775  2016-12-31 15:59:12          Canada      2203  CAN6    Male   
14968     65776  2016-12-31 17:54:24         Germany      2231  GER1  Female   
14969     65777  2016-12-31 18:52:00         Germany      2156  GER1  Female   

      Size (US) Size (Europe) Size (UK) UnitPrice Discount SalePrice  
14965       9.5         42-43         9       139        0       139  
14966        12         42-43        10       149        0       149  
14967      10.5         43-44        10       179      0.3     125.3  
14968       9.5            40       7.5       199        0       199  
14969       6.5            37       4.5       139      0.1     125.1

In [471]:
df['Date'] = df['Date'].astype(str).str[:7]
df

InvoiceNo     Date         Country ProductID  Shop  Gender Size (US)  \
3         52389  2014-01  United Kingdom      2152   UK2    Male        11   
4         52390  2014-01   United States      2230  US15    Male      11.5   
5         52391  2014-01          Canada      2160  CAN7    Male       9.5   
6         52392  2014-01   United States      2234   US6  Female       9.5   
7         52393  2014-01  United Kingdom      2222   UK4  Female         9   
...         ...      ...             ...       ...   ...     ...       ...   
14965     65773  2016-12  United Kingdom      2154   UK2    Male       9.5   
14966     65774  2016-12   United States      2181  US12  Female        12   
14967     65775  2016-12          Canada      2203  CAN6    Male      10.5   
14968     65776  2016-12         Germany      2231  GER1  Female       9.5   
14969     65777  2016-12         Germany      2156  GER1  Female       6.5   

      Size (Europe) Size (UK) UnitPrice Discount SalePrice  
3                44      10.5       159        0       159  
4             44-45        11       199      0.2     159.2  
5             42-43         9       149      0.2     119.2  
6                40       7.5       159        0       159  
7             39-40         7       159        0       159  
...             ...       ...       ...      ...       ...  
14965         42-43         9       139        0       139  
14966         42-43        10       149        0       149  
14967         43-44        10       179      0.3     125.3  
14968            40       7.5       199        0       199  
14969            37       4.5       139      0.1     125.1  

[14967 rows x 12 columns]

In [472]:
df.isnull().sum()

InvoiceNo        0
Date             0
Country          0
ProductID        0
Shop             0
Gender           0
Size (US)        0
Size (Europe)    0
Size (UK)        0
UnitPrice        0
Discount         0
SalePrice        0
dtype: int64

In [473]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14967 entries, 3 to 14969
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   InvoiceNo      14967 non-null  object
 1   Date           14967 non-null  object
 2   Country        14967 non-null  object
 3   ProductID      14967 non-null  object
 4   Shop           14967 non-null  object
 5   Gender         14967 non-null  object
 6   Size (US)      14967 non-null  object
 7   Size (Europe)  14967 non-null  object
 8   Size (UK)      14967 non-null  object
 9   UnitPrice      14967 non-null  object
 10  Discount       14967 non-null  object
 11  SalePrice      14967 non-null  object
dtypes: object(12)
memory usage: 1.4+ MB


In [474]:
df.columns

Index(['InvoiceNo', 'Date', 'Country', 'ProductID', 'Shop', 'Gender',
       'Size (US)', 'Size (Europe)', 'Size (UK)', 'UnitPrice', 'Discount',
       'SalePrice'],
      dtype='object')

In [475]:
df_men_usa = df[['Date','Gender','Country', 'Size (US)']].copy()
df_men_usa

Date  Gender         Country Size (US)
3      2014-01    Male  United Kingdom        11
4      2014-01    Male   United States      11.5
5      2014-01    Male          Canada       9.5
6      2014-01  Female   United States       9.5
7      2014-01  Female  United Kingdom         9
...        ...     ...             ...       ...
14965  2016-12    Male  United Kingdom       9.5
14966  2016-12  Female   United States        12
14967  2016-12    Male          Canada      10.5
14968  2016-12  Female         Germany       9.5
14969  2016-12  Female         Germany       6.5

[14967 rows x 4 columns]

In [476]:
df_men_usa = df_men_usa[df_men_usa.Gender != 'Female']
df_men_usa = df_men_usa[df_men_usa.Country == 'United States']
df_men_usa = df_men_usa[df_men_usa.Date >= '2015-01']
df_men_usa = df_men_usa[df_men_usa.Date <= '2016-12']

df_men_usa.head(660)

Date Gender        Country Size (US)
2756  2015-01   Male  United States       9.5
2757  2015-01   Male  United States       8.5
2758  2015-01   Male  United States         9
2769  2015-01   Male  United States        10
2770  2015-01   Male  United States       9.5
...       ...    ...            ...       ...
5693  2015-08   Male  United States         9
5698  2015-08   Male  United States        10
5700  2015-08   Male  United States       9.5
5702  2015-08   Male  United States        11
5707  2015-08   Male  United States         6

[660 rows x 4 columns]

In [477]:
df_groupby_size_date = df_men_usa.groupby(['Date', 'Size (US)']).agg({'Gender': pd.Series.count })
df_groupby_size_date.head(20)

Gender
Date    Size (US)        
2015-01 6.5             4
        8.0             5
        8.5             4
        9.0            10
        9.5            18
        10.0            8
        10.5            7
        11.0            5
        11.5            3
        12.0            1
        14.0            1
2015-02 6.5             1
        7.5             1
        8.0             3
        8.5             5
        9.0            14
        9.5            16
        10.0           13
        10.5           13
        11.0            5

In [478]:
df_groupby_size_date_2 = df_groupby_size_date.groupby(['Size (US)']).agg({'Gender': pd.Series.sum }) 
meanX = df_groupby_size_date_2 / 24
meanX.round(2)

Gender
Size (US)        
6.0          2.17
6.5          1.58
7.0          1.33
7.5          2.33
8.0          4.79
8.5          7.88
9.0         16.33
9.5         25.58
10.0        18.79
10.5        14.96
11.0         7.54
11.5         5.33
12.0         3.08
13.0         1.21
14.0         1.96
15.0         0.54

In [479]:
pre_variance = (df_groupby_size_date - meanX)**2
variance  = pre_variance.groupby('Size (US)').agg({'Gender': pd.Series.sum}) / 24 
variance

Gender
Size (US)           
6.0         2.186343
6.5         1.928530
7.0         1.824074
7.5         2.981481
8.0         6.334925
8.5        20.526042
9.0        36.638889
9.5        71.743056
10.0       40.414931
10.5       23.956597
11.0       11.914931
11.5        7.805556
12.0        6.158565
13.0        0.950738
14.0        1.560547
15.0        0.648220

In [480]:
std = variance.Gender ** 0.5
std

Size (US)
6.0     1.478629
6.5     1.388715
7.0     1.350583
7.5     1.726697
8.0     2.516928
8.5     4.530567
9.0     6.053007
9.5     8.470127
10.0    6.357274
10.5    4.894548
11.0    3.451801
11.5    2.793842
12.0    2.481646
13.0    0.975058
14.0    1.249219
15.0    0.805121
Name: Gender, dtype: float64

In [481]:
std_error = std / math.sqrt(24)
std_error

Size (US)
6.0     0.301824
6.5     0.283470
7.0     0.275687
7.5     0.352460
8.0     0.513766
8.5     0.924798
9.0     1.235565
9.5     1.728957
10.0    1.297673
10.5    0.999095
11.0    0.704596
11.5    0.570291
12.0    0.506564
13.0    0.199033
14.0    0.254996
15.0    0.164345
Name: Gender, dtype: float64

COFIDENCE INTERVALS

In [482]:
T = 2.64
ci_a = meanX.Gender + (T * std_error) 
ci_a.round(2)

Size (US)
6.0      2.96
6.5      2.33
7.0      2.06
7.5      3.26
8.0      6.15
8.5     10.32
9.0     19.60
9.5     30.15
10.0    22.22
10.5    17.60
11.0     9.40
11.5     6.84
12.0     4.42
13.0     1.73
14.0     2.63
15.0     0.98
Name: Gender, dtype: float64

In [483]:
ci_b = meanX.Gender - (T * std_error) 
ci_b.round(2)

Size (US)
6.0      1.37
6.5      0.83
7.0      0.61
7.5      1.40
8.0      3.44
8.5      5.43
9.0     13.07
9.5     21.02
10.0    15.37
10.5    12.32
11.0     5.68
11.5     3.83
12.0     1.75
13.0     0.68
14.0     1.29
15.0     0.11
Name: Gender, dtype: float64

### Number of pairs we need in stock per store, math round  2015-2016<br>
Size (US)<br>
6.0 = 3<br>
6.5 = 2<br>
7.0 = 2<br>
7.5 = 3<br>
8.0 = 6<br>
8.5 = 10<br>
9.0 = 20<br>
9.5 = 30<br>
10.0 = 22<br>
10.5 = 18<br>
11.0 = 9<br>
11.5 = 7<br>
12.0 = 4<br>
13.0 = 2<br>
14.0 = 3<br>
15.0 = 1<br>
### Number of pairs we need in stock per store, math round  2016<br>
Size (US)<br>
6.0 = 4<br>
6.5 = 3<br>
7.0 = 3<br>
7.5 = 5<br>
8.0 = 8<br>
8.5 = 14<br>
9.0 = 24<br>
9.5 = 37<br>
10.0 = 27<br>
10.5 = 21<br>
11.0 = 12<br>
11.5 = 8<br>
12.0 = 6<br>
13.0 = 2<br>
14.0 = 4<br>
15.0 = 1<br>

In [484]:
test = [0, 0, 0, 0, 3, 1, 1, 3, 5, 4, 0, 0, 4, 1, 3, 1, 3, 3, 3, 4, 3, 7, 3, 0]
x = []
y = 0
z = 0
for i in test :
    y = (i - 2.17) ** 2
    x.append(y)
for i in x:
    z += i 
h_variance = z / 24    
h_variance
std_h = math.sqrt(h_variance)   
std_h
std_error_h = std_h / math.sqrt(24)
std_error_h

0.3849007808658111